### PR TITLE
Added example for integrating a0+fbrp to IFTTT

### LINKED ISSUE #876: [fbrp] IFTTT
Provide tooling or examples of how to trigger IFTTT.

Select the type of issue:
- [ ] Bug report (to report a bug)
- [x] Feature request (to request an additional feature)
- [x] Tracker (I am just using this as a tracker)
- [ ] Refactor request
- [ ] Documentation Ask

### PR DESCRIPTION
**Pull request by zephirefaith**
# Description

An example/tutorial to exhibit integration of IFTTT Webhook applets into a0 + fbrp environment to enable arbitrary email/online alerts. The ifttt client is completely configurable and the example includes instructions on how can one use it.

Fixes #876 

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

# Testing

Followed the instructions in example with my IFTTT_KEY and verified emails being received.

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [x] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

### CODE REVIEW

File: fbrp/examples/10_ifttt/ifttt.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+
+def call_robo_crit_event(pkt):
+    print(f"Pinging IFTTT Event {EVENT_NAME}")
+    requests.post(f"https://maker.ifttt.com/trigger/{EVENT_NAME}/with/key/{KEY}")
+
+
+s = a0.Subscriber(f"{EVENT_TOPIC}", lambda pkt: call_robo_crit_event(pkt))
+
+# a0.update_configs()
```
Comment (lshamis) [inline fbrp/examples/10_ifttt/ifttt.py:18]: Remove dead code.
If you want to include `a0.update_configs()` somewhere, put it into the `call_robo_crit_event` function.

File: fbrp/examples/10_ifttt/ifttt.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+EVENT_TOPIC = a0.cfg(a0.env.topic(), "/event_topic", str)
+a0.update_configs()
+
+
+def call_robo_crit_event(pkt):
+    print(f"Pinging IFTTT Event {EVENT_NAME}")
+    requests.post(f"https://maker.ifttt.com/trigger/{EVENT_NAME}/with/key/{KEY}")
+
+
+s = a0.Subscriber(f"{EVENT_TOPIC}", lambda pkt: call_robo_crit_event(pkt))
```
Comment (lshamis) [inline fbrp/examples/10_ifttt/ifttt.py:16]: Change
```py
lambda pkt: call_robo_crit_event(pkt)
```
to
```py
call_robo_crit_event
```

File: fbrp/examples/10_ifttt/ifttt.py (hunk 1 above)
Comment (lshamis) [reply fbrp/examples/10_ifttt/ifttt.py:18]: Note: It wont change the EVENT_TOPIC used by the Subscriber.

**Review by lshamis — CHANGES_REQUESTED** [review]

File: fbrp/examples/10_ifttt/fsetup.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+fbrp.process(
+    name="ifttt_webhook",
+    runtime=fbrp.Conda(
+        yaml="env.yml",
+        run_command=["python3", "ifttt.py"],
+    ),
+    cfg={
+        "ifttt": {"key": f"{IFTTT_KEY}", "event_name": f"{TRIGGER_EVENT_NAME}"},
+        "event_topic": f"{TRIGGER_EVENT_TOPIC}",
```
Comment (lshamis) [inline fbrp/examples/10_ifttt/fsetup.py:26]: I don't think you need all the `f"{...}"`
`f"{TRIGGER_EVENT_TOPIC}"` can just be `TRIGGER_EVENT_TOPIC`

File: fbrp/examples/10_ifttt/fsetup.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    },
+)
+
+fbrp.process(
+    name="api",
```
Comment (lshamis) [inline fbrp/examples/10_ifttt/fsetup.py:31]: You can probably remove `api` here. I don't think it's useful for the example.

File: fbrp/examples/10_ifttt/ifttt.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+EVENT_NAME = a0.cfg(a0.env.topic(), "/ifttt/event_name", str)
+KEY = a0.cfg(a0.env.topic(), "/ifttt/key", str)
+EVENT_TOPIC = a0.cfg(a0.env.topic(), "/event_topic", str)
+a0.update_configs()
+
+
+def call_robo_crit_event(pkt):
+    print(f"Pinging IFTTT Event {EVENT_NAME}")
+    requests.post(f"https://maker.ifttt.com/trigger/{EVENT_NAME}/with/key/{KEY}")
```
Comment (lshamis) [inline fbrp/examples/10_ifttt/ifttt.py:13]: Not necessarily in this PR, but do you want to post the message content?

File: fbrp/examples/10_ifttt/ifttt.py (hunk 1 above)
Comment (zephirefaith) [reply fbrp/examples/10_ifttt/ifttt.py:18]: Thanks for pointing this out: Wouldn't we want `s` to change subscription though if we are permitting online update of parameters? Is there a better way of doing this?

File: fbrp/examples/10_ifttt/ifttt.py (hunk 5 above)
Comment (zephirefaith) [reply fbrp/examples/10_ifttt/ifttt.py:13]: It's pretty straightforward to include the message content as well, I can just extend the current one and add comment on how to use if the user does not want the message.

File: fbrp/examples/10_ifttt/ifttt.py (hunk 1 above)
Comment (lshamis) [reply fbrp/examples/10_ifttt/ifttt.py:18]: Uhh... ok. I don't necessarily think we should support changing all the variables. I think the code complexity will increase faster than the usefulness.

The following is an example of how you would do it, if you really wanted:
```py
class Foo:
    def __init__(self):
        self._topic: str = None
        self._s: a0.Subscriber = None

    def on_cfg_change(self, pkt):
        cfg = json.loads(pkt.payload)
        if self._topic != cfg.get("event_topic"):
            self._topic = cfg.get("event_topic")
            self._s = None
            if self._topic:
                self._s = a0.Subscriber(self._topic, call_robo_crit_event)


watcher = a0.CfgWatcher(a0.env.topic(), Foo().on_cfg_change)
```

`CfgWatcher` is not smart enough to watch a sub-field.
None of the code here was tested, at all.

File: fbrp/examples/10_ifttt/ifttt.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import a0
+import signal
+import requests
+
+# read in the parameters from fsetup cfg
+EVENT_NAME = a0.cfg(a0.env.topic(), "/ifttt/event_name", str)
+KEY = a0.cfg(a0.env.topic(), "/ifttt/key", str)
+EVENT_TOPIC = a0.cfg(a0.env.topic(), "/event_topic", str)
+a0.update_configs()
```
Comment (lshamis) [inline fbrp/examples/10_ifttt/ifttt.py:9]: This update doesn't do anything.

File: fbrp/examples/10_ifttt/ifttt.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+
+def call_robo_crit_event(topic_name, pkt):
```
Comment (lshamis) [inline fbrp/examples/10_ifttt/ifttt.py:12]: You've already made the topic_name global. Just use EVENT_TOPIC directly.

File: fbrp/examples/10_ifttt/trigger_event_detector.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+print(f"{EVENT_TOPIC} initialized")
+print(type(EVENT_TOPIC))
+
+i = 0
+p = a0.Publisher(f"{EVENT_TOPIC}")
+print(f"we have a publisher")
+
+while True:
+    print(f"updating config")
+    a0.update_configs()
```
Comment (lshamis) [inline fbrp/examples/10_ifttt/trigger_event_detector.py:14]: What does this update?
There are no cfgs used in this loop. You can drop this line.

File: fbrp/examples/10_ifttt/trigger_event_detector.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    msg = f"data {i}"
+    print(f"stdout {msg}")
```
Comment (lshamis) [inline fbrp/examples/10_ifttt/trigger_event_detector.py:16]:
```py
print(msg)
```

File: fbrp/examples/10_ifttt/ifttt.py [hunk 10]
```diff
@@ -27,7 +26,7 @@ def call_robo_crit_event(topic_name, pkt):
 
 s = a0.Subscriber(
     f"{EVENT_TOPIC}",
-    lambda pkt: call_robo_crit_event(f"{EVENT_TOPIC}", pkt),
+    lambda pkt: call_robo_crit_event(pkt),
```
Comment (lshamis) [inline fbrp/examples/10_ifttt/ifttt.py:29]: Change
```py
lambda pkt: call_robo_crit_event(pkt),
```
to
```py
call_robo_crit_event
```

The lambda doesn't add anything

**Review by lshamis — APPROVED** [review]

**Pull request merged**